// @tweakpane/core/dist/es6/blade/rack/controller/rack.js
import { insertElementAt, removeElement } from '../../../common/dom-util';
import { PlainView } from '../../../common/view/plain';
import { BladeController } from '../../common/controller/blade';
import { BladeRack } from '../../common/model/blade-rack';
export class RackController extends BladeController {
    constructor(doc, config) {
        super(Object.assign(Object.assign({}, config), { view: new PlainView(doc, {
                viewName: 'brk',
                viewProps: config.viewProps,
            }) }));
        this.onRackAdd_ = this.onRackAdd_.bind(this);
        this.onRackRemove_ = this.onRackRemove_.bind(this);
        const rack = new BladeRack({
            blade: config.root ? undefined : config.blade,
            viewProps: config.viewProps,
        });
        rack.emitter.on('add', this.onRackAdd_);
        rack.emitter.on('remove', this.onRackRemove_);
        this.rack = rack;
        this.viewProps.handleDispose(() => {
            for (let i = this.rack.children.length - 1; i >= 0; i--) {
                const bc = this.rack.children[i];
                bc.viewProps.set('disposed', true);
            }
        });
    }
    onRackAdd_(ev) {
        if (!ev.isRoot) {
            return;
        }
        insertElementAt(this.view.element, ev.bladeController.view.element, ev.index);
    }
    onRackRemove_(ev) {
        if (!ev.isRoot) {
            return;
        }
        removeElement(ev.bladeController.view.element);
    }
}
