// @tweakpane/core/dist/es6/blade/common/api/tp-event.js
/**
 * A base class of Tweakpane API events.
 */
export class TpEvent {
    /**
     * @hidden
     */
    constructor(target) {
        this.target = target;
    }
}
/**
 * An event class for value changes of input bindings.
 * @template T The type of the value.
 */
export class TpChangeEvent extends TpEvent {
    /**
     * @hidden
     */
    constructor(target, value, presetKey, last) {
        super(target);
        this.value = value;
        this.presetKey = presetKey;
        this.last = last !== null && last !== void 0 ? last : true;
    }
}
/**
 * An event class for value updates of monitor bindings.
 * @template T The type of the value.
 */
export class TpUpdateEvent extends TpEvent {
    /**
     * @hidden
     */
    constructor(target, value, presetKey) {
        super(target);
        this.value = value;
        this.presetKey = presetKey;
    }
}
/**
 * An event class for folder.
 */
export class TpFoldEvent extends TpEvent {
    /**
     * @hidden
     */
    constructor(target, expanded) {
        super(target);
        this.expanded = expanded;
    }
}
/**
 * An event class for tab selection.
 */
export class TpTabSelectEvent extends TpEvent {
    /**
     * @hidden
     */
    constructor(target, index) {
        super(target);
        this.index = index;
    }
}
