// @tweakpane/core/dist/es6/blade/monitor-binding/api/monitor-binding.js
import { Emitter } from '../../../common/model/emitter';
import { forceCast } from '../../../misc/type-util';
import { BladeApi } from '../../common/api/blade';
import { TpUpdateEvent } from '../../common/api/tp-event';
/**
 * The API for the monitor binding between the parameter and the pane.
 */
export class MonitorBindingApi extends BladeApi {
    /**
     * @hidden
     */
    constructor(controller) {
        super(controller);
        this.onBindingUpdate_ = this.onBindingUpdate_.bind(this);
        this.emitter_ = new Emitter();
        this.controller_.binding.emitter.on('update', this.onBindingUpdate_);
    }
    get label() {
        return this.controller_.props.get('label');
    }
    set label(label) {
        this.controller_.props.set('label', label);
    }
    on(eventName, handler) {
        const bh = handler.bind(this);
        this.emitter_.on(eventName, (ev) => {
            bh(ev.event);
        });
        return this;
    }
    refresh() {
        this.controller_.binding.read();
    }
    onBindingUpdate_(ev) {
        const value = ev.sender.target.read();
        this.emitter_.emit('update', {
            event: new TpUpdateEvent(this, forceCast(value), this.controller_.binding.target.presetKey),
        });
    }
}
