// @tweakpane/core/dist/es6/blade/common/controller/blade.js
import { warnMissing } from '../../../common/compat';
import { disposeElement } from '../../../common/disposing-util';
import { ViewProps } from '../../../common/model/view-props';
import { ClassName } from '../../../common/view/class-name';
import { getAllBladePositions } from '../model/blade-positions';
const className = ClassName('');
const POS_TO_CLASS_NAME_MAP = {
    veryfirst: 'vfst',
    first: 'fst',
    last: 'lst',
    verylast: 'vlst',
};
export class BladeController {
    constructor(config) {
        this.parent_ = null;
        this.blade = config.blade;
        this.view = config.view;
        this.viewProps = config.viewProps;
        const elem = this.view.element;
        this.blade.value('positions').emitter.on('change', () => {
            getAllBladePositions().forEach((pos) => {
                elem.classList.remove(className(undefined, POS_TO_CLASS_NAME_MAP[pos]));
            });
            this.blade.get('positions').forEach((pos) => {
                elem.classList.add(className(undefined, POS_TO_CLASS_NAME_MAP[pos]));
            });
        });
        this.viewProps.handleDispose(() => {
            disposeElement(elem);
        });
    }
    get parent() {
        return this.parent_;
    }
    set parent(parent) {
        this.parent_ = parent;
        // TODO: Remove it in the next major version
        if (!('parent' in this.viewProps.valMap_)) {
            warnMissing({
                key: 'parent',
                target: ViewProps.name,
                place: 'BladeController.parent',
            });
            return;
        }
        this.viewProps.set('parent', this.parent_ ? this.parent_.viewProps : null);
    }
}
