// @tweakpane/core/dist/es6/input-binding/string/plugin.js
import { CompositeConstraint, findConstraint, } from '../../common/constraint/composite';
import { ListConstraint } from '../../common/constraint/list';
import { ListController } from '../../common/controller/list';
import { TextController } from '../../common/controller/text';
import { formatString, stringFromUnknown } from '../../common/converter/string';
import { ValueMap } from '../../common/model/value-map';
import { ParamsParsers, parseParams } from '../../common/params-parsers';
import { writePrimitive } from '../../common/primitive';
import { createListConstraint, parseListOptions } from '../../common/util';
function createConstraint(params) {
    const constraints = [];
    const lc = createListConstraint(params.options);
    if (lc) {
        constraints.push(lc);
    }
    return new CompositeConstraint(constraints);
}
/**
 * @hidden
 */
export const StringInputPlugin = {
    id: 'input-string',
    type: 'input',
    accept: (value, params) => {
        if (typeof value !== 'string') {
            return null;
        }
        const p = ParamsParsers;
        const result = parseParams(params, {
            options: p.optional.custom(parseListOptions),
        });
        return result
            ? {
                initialValue: value,
                params: result,
            }
            : null;
    },
    binding: {
        reader: (_args) => stringFromUnknown,
        constraint: (args) => createConstraint(args.params),
        writer: (_args) => writePrimitive,
    },
    controller: (args) => {
        const doc = args.document;
        const value = args.value;
        const c = args.constraint;
        const lc = c && findConstraint(c, ListConstraint);
        if (lc) {
            return new ListController(doc, {
                props: new ValueMap({
                    options: lc.values.value('options'),
                }),
                value: value,
                viewProps: args.viewProps,
            });
        }
        return new TextController(doc, {
            parser: (v) => v,
            props: ValueMap.fromObject({
                formatter: formatString,
            }),
            value: value,
            viewProps: args.viewProps,
        });
    },
};
