// @tweakpane/core/dist/es6/input-binding/color/view/color-text.js
import { createSvgIconElement, removeChildElements, } from '../../../common/dom-util';
import { bindValue } from '../../../common/model/reactive';
import { ClassName } from '../../../common/view/class-name';
const className = ClassName('coltxt');
function createModeSelectElement(doc) {
    const selectElem = doc.createElement('select');
    const items = [
        { text: 'RGB', value: 'rgb' },
        { text: 'HSL', value: 'hsl' },
        { text: 'HSV', value: 'hsv' },
    ];
    selectElem.appendChild(items.reduce((frag, item) => {
        const optElem = doc.createElement('option');
        optElem.textContent = item.text;
        optElem.value = item.value;
        frag.appendChild(optElem);
        return frag;
    }, doc.createDocumentFragment()));
    return selectElem;
}
/**
 * @hidden
 */
export class ColorTextView {
    constructor(doc, config) {
        this.element = doc.createElement('div');
        this.element.classList.add(className());
        config.viewProps.bindClassModifiers(this.element);
        const modeElem = doc.createElement('div');
        modeElem.classList.add(className('m'));
        this.modeElem_ = createModeSelectElement(doc);
        this.modeElem_.classList.add(className('ms'));
        modeElem.appendChild(this.modeSelectElement);
        config.viewProps.bindDisabled(this.modeElem_);
        const modeMarkerElem = doc.createElement('div');
        modeMarkerElem.classList.add(className('mm'));
        modeMarkerElem.appendChild(createSvgIconElement(doc, 'dropdown'));
        modeElem.appendChild(modeMarkerElem);
        this.element.appendChild(modeElem);
        const textsElem = doc.createElement('div');
        textsElem.classList.add(className('w'));
        this.element.appendChild(textsElem);
        this.textsElem_ = textsElem;
        this.textViews_ = config.textViews;
        this.applyTextViews_();
        bindValue(config.colorMode, (mode) => {
            this.modeElem_.value = mode;
        });
    }
    get modeSelectElement() {
        return this.modeElem_;
    }
    get textViews() {
        return this.textViews_;
    }
    set textViews(textViews) {
        this.textViews_ = textViews;
        this.applyTextViews_();
    }
    applyTextViews_() {
        removeChildElements(this.textsElem_);
        const doc = this.element.ownerDocument;
        this.textViews_.forEach((v) => {
            const compElem = doc.createElement('div');
            compElem.classList.add(className('c'));
            compElem.appendChild(v.element);
            this.textsElem_.appendChild(compElem);
        });
    }
}
