// @tweakpane/core/dist/es6/common/model/emitter.js
/**
 * A type-safe event emitter.
 * @template E The interface that maps event names and event objects.
 */
export class Emitter {
    constructor() {
        this.observers_ = {};
    }
    /**
     * Adds an event listener to the emitter.
     * @param eventName The event name to listen.
     * @param handler The event handler.
     */
    on(eventName, handler) {
        let observers = this.observers_[eventName];
        if (!observers) {
            observers = this.observers_[eventName] = [];
        }
        observers.push({
            handler: handler,
        });
        return this;
    }
    /**
     * Removes an event listener from the emitter.
     * @param eventName The event name.
     * @param handler The event handler to remove.
     */
    off(eventName, handler) {
        const observers = this.observers_[eventName];
        if (observers) {
            this.observers_[eventName] = observers.filter((observer) => {
                return observer.handler !== handler;
            });
        }
        return this;
    }
    emit(eventName, event) {
        const observers = this.observers_[eventName];
        if (!observers) {
            return;
        }
        observers.forEach((observer) => {
            observer.handler(event);
        });
    }
}
