// @tweakpane/core/dist/es6/common/params-parsers.js
import { forceCast } from '../misc/type-util';
function parseObject(value, keyToParserMap) {
    const keys = Object.keys(keyToParserMap);
    const result = keys.reduce((tmp, key) => {
        if (tmp === undefined) {
            return undefined;
        }
        const parser = keyToParserMap[key];
        const result = parser(value[key]);
        return result.succeeded
            ? Object.assign(Object.assign({}, tmp), { [key]: result.value }) : undefined;
    }, {});
    return forceCast(result);
}
function parseArray(value, parseItem) {
    return value.reduce((tmp, item) => {
        if (tmp === undefined) {
            return undefined;
        }
        const result = parseItem(item);
        if (!result.succeeded || result.value === undefined) {
            return undefined;
        }
        return [...tmp, result.value];
    }, []);
}
function isObject(value) {
    if (value === null) {
        return false;
    }
    return typeof value === 'object';
}
function createParamsParserBuilder(parse) {
    return (optional) => (v) => {
        if (!optional && v === undefined) {
            return {
                succeeded: false,
                value: undefined,
            };
        }
        if (optional && v === undefined) {
            return {
                succeeded: true,
                value: undefined,
            };
        }
        const result = parse(v);
        return result !== undefined
            ? {
                succeeded: true,
                value: result,
            }
            : {
                succeeded: false,
                value: undefined,
            };
    };
}
function createParamsParserBuilders(optional) {
    return {
        custom: (parse) => createParamsParserBuilder(parse)(optional),
        boolean: createParamsParserBuilder((v) => typeof v === 'boolean' ? v : undefined)(optional),
        number: createParamsParserBuilder((v) => typeof v === 'number' ? v : undefined)(optional),
        string: createParamsParserBuilder((v) => typeof v === 'string' ? v : undefined)(optional),
        // eslint-disable-next-line @typescript-eslint/ban-types
        function: createParamsParserBuilder((v) => 
        // eslint-disable-next-line @typescript-eslint/ban-types
        typeof v === 'function' ? v : undefined)(optional),
        constant: (value) => createParamsParserBuilder((v) => (v === value ? value : undefined))(optional),
        raw: createParamsParserBuilder((v) => v)(optional),
        object: (keyToParserMap) => createParamsParserBuilder((v) => {
            if (!isObject(v)) {
                return undefined;
            }
            return parseObject(v, keyToParserMap);
        })(optional),
        array: (itemParser) => createParamsParserBuilder((v) => {
            if (!Array.isArray(v)) {
                return undefined;
            }
            return parseArray(v, itemParser);
        })(optional),
    };
}
export const ParamsParsers = {
    optional: createParamsParserBuilders(true),
    required: createParamsParserBuilders(false),
};
export function parseParams(value, keyToParserMap) {
    const result = ParamsParsers.required.object(keyToParserMap)(value);
    return result.succeeded ? result.value : undefined;
}
