// @tweakpane/core/dist/es6/blade/folder/plugin.js
import { ValueMap } from '../../common/model/value-map';
import { ParamsParsers, parseParams } from '../../common/params-parsers';
import { FolderApi } from './api/folder';
import { FolderController } from './controller/folder';
export const FolderBladePlugin = {
    id: 'folder',
    type: 'blade',
    accept(params) {
        const p = ParamsParsers;
        const result = parseParams(params, {
            title: p.required.string,
            view: p.required.constant('folder'),
            expanded: p.optional.boolean,
        });
        return result ? { params: result } : null;
    },
    controller(args) {
        return new FolderController(args.document, {
            blade: args.blade,
            expanded: args.params.expanded,
            props: ValueMap.fromObject({
                title: args.params.title,
            }),
            viewProps: args.viewProps,
        });
    },
    api(args) {
        if (!(args.controller instanceof FolderController)) {
            return null;
        }
        return new FolderApi(args.controller, args.pool);
    },
};
