// @tweakpane/core/dist/es6/blade/tab/api/tab-page.js
export class TabPageApi {
    constructor(controller, contentRackApi) {
        this.controller_ = controller;
        this.rackApi_ = contentRackApi;
    }
    get title() {
        var _a;
        return (_a = this.controller_.itemController.props.get('title')) !== null && _a !== void 0 ? _a : '';
    }
    set title(title) {
        this.controller_.itemController.props.set('title', title);
    }
    get selected() {
        return this.controller_.props.get('selected');
    }
    set selected(selected) {
        this.controller_.props.set('selected', selected);
    }
    get children() {
        return this.rackApi_.children;
    }
    addButton(params) {
        return this.rackApi_.addButton(params);
    }
    addFolder(params) {
        return this.rackApi_.addFolder(params);
    }
    addSeparator(opt_params) {
        return this.rackApi_.addSeparator(opt_params);
    }
    addTab(params) {
        return this.rackApi_.addTab(params);
    }
    add(api, opt_index) {
        this.rackApi_.add(api, opt_index);
    }
    remove(api) {
        this.rackApi_.remove(api);
    }
    addInput(object, key, opt_params) {
        return this.rackApi_.addInput(object, key, opt_params);
    }
    addMonitor(object, key, opt_params) {
        return this.rackApi_.addMonitor(object, key, opt_params);
    }
    addBlade(params) {
        return this.rackApi_.addBlade(params);
    }
}
