// @tweakpane/core/dist/es6/monitor-binding/boolean/plugin.js
import { BooleanFormatter, boolFromUnknown, } from '../../common/converter/boolean';
import { ParamsParsers, parseParams } from '../../common/params-parsers';
import { Constants } from '../../misc/constants';
import { MultiLogController } from '../common/controller/multi-log';
import { SingleLogController } from '../common/controller/single-log';
/**
 * @hidden
 */
export const BooleanMonitorPlugin = {
    id: 'monitor-bool',
    type: 'monitor',
    accept: (value, params) => {
        if (typeof value !== 'boolean') {
            return null;
        }
        const p = ParamsParsers;
        const result = parseParams(params, {
            lineCount: p.optional.number,
        });
        return result
            ? {
                initialValue: value,
                params: result,
            }
            : null;
    },
    binding: {
        reader: (_args) => boolFromUnknown,
    },
    controller: (args) => {
        var _a;
        if (args.value.rawValue.length === 1) {
            return new SingleLogController(args.document, {
                formatter: BooleanFormatter,
                value: args.value,
                viewProps: args.viewProps,
            });
        }
        return new MultiLogController(args.document, {
            formatter: BooleanFormatter,
            lineCount: (_a = args.params.lineCount) !== null && _a !== void 0 ? _a : Constants.monitor.defaultLineCount,
            value: args.value,
            viewProps: args.viewProps,
        });
    },
};
